csFastFloat/FastDoubleParser.cs

```diff
@@ -412,6 +454,45 @@ unsafe static internal double HandleInvalidInput(char* first, char* last)
     }
 
 
+    unsafe static internal double HandleInvalidInput(byte* first, byte* last)
+    {
+      ReadOnlySpan<byte> infinity_string = new byte[]{105, 110, 102, 105, 110, 105, 116, 121};
+      ReadOnlySpan<byte> inf_string = new byte[]{105, 110, 102};
```

Comment:
Just FYI as this will be out of scope here. For utf-8 based string literals there's a nice project --> https://github.com/ufcpp/StringLiteralGenerator
It uses C# 9's source generators to create the byte-sequence for us.

Reply:
Ok. As you may imagine, I felt a bit desperate... 

I take it that is no easy way to express UTF-8 string literals in C#, right? Like `"my string"_utf8` or something? (C++ allows that.)

Reply:
UTF-8 literals will be supported by a future version of C#.

Reply:
While it's probably even uglier, I prefer the following myself::
```
ReadOnlySpan<byte> inf_string = new byte[]{(byte)'i', (byte)'n', (byte)'f'};
```
At least a casual viewer can read that and know that this isn't a bunch of magic numbers.

Reply:
@mburbea @gfoidl I guess that there is no chance that `System.Text.Encoding.UTF8.GetBytes` would be resolved at compile time?

Reply:
I have added a detailed description of the arrays.

Reply:
> no chance that `System.Text.Encoding.UTF8.GetBytes` would be resolved at compile time?

Concerning C#: no -- only with the source generator (see comment above).
Concerning JIT: yes.

Before writing something about the JIT-variant, I'd like to state that the source generator variant (C#) should be done, as it's the cleanest and easiest to read one. I can put up a PR (to your branch) if you want.

The JIT-case for optimal performance isn't very intuitive though.

`static readonly` fields are treated by JIT as "constant". According the ECMA-rules and `beforefieldinit` IL-Instruction, which is emitted by C# for the static readonly, the value must be initialized before the first access. So in order for the runtime to know when it's time to initialize the JIT has to emit a check for this, and as you can imagine this check has a tiny cost.

Now the JIT is tiered, which means at startup methods are compiled with less optimizations at so called Tier-0. This allows faster start-times. If a method is called several times (current threshould is 30) after a startup-period (current: 100 ms), it's re-JITed with optimization to Tier-1 code.
At this re-JITing it's guaranteed that static readonly fields are already initialized, thus at Tier-1 the beforementioned check isn't emitted.

But there's a caveat: some methods with loops got trapped in Tier-0 as the call-count didn't increase, so the less optimized version of the code ran (for ever). Hence the JIT-team decided to create Tier-1 code for methods with loops.
For the static readonly case this means, that the check for initialization must be emitted.

It wouldn't be .NET if there weren't any trick to elide this check and have Tiere-1 code for loops that are performant.
Among other options like config-tweaks for tiered compilation, a `ModuleInitializer` (new C# 9 feature) can be used. This ModuleInitializer is guaranteed to run before any other code in that module, so if this ModuleInitializer accesses the static readonly field, the JIT knows it's initialized, so the check can be omitted. https://gist.github.com/gfoidl/45d1f4b47fae81f49656d4c52a169ac9 demonstrates that.

Reply:
https://github.com/CarlVerret/csFastFloat/pull/60 brings in the source generator approach.